### PR TITLE
Setup JT4 Networking

### PR DESCRIPTION
**Pull request by jumaallan**
Story: https://app.shortcut.com/smileid/story/12098/replace-prep-upload-with-new-fast-jt4-endpoint-for-biometric-enrollment-v10-ios

## Summary

Replace Prep Upload with new fast JT4 endpoint for Biometric enrollment on iOS

## Known Issues

Any shortcomings in your work. This may include corner cases not correctly handled or issues related
to but not within the scope of your PR. Design compromises should be discussed here if they were not
already discussed above.

## Test Instructions

Concise test instructions on how to verify that your feature works as intended. This should include
changes to the development environment (if applicable) and all commands needed to run your work.

## Screenshot

If applicable (e.g. UI changes), add screenshots to help explain your work.

### CODE REVIEW

File: Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift [hunk 1]
```diff
@@ -314,20 +305,53 @@ public class SelfieViewModel: ObservableObject, ARKitSmileDelegate {
...
+                        }
+                        return nil
+                    }
+
+                    imageInfoArray.append(contentsOf: livenessImageInfos.compactMap { $0 })
+                }
+                let response = if allowNewEnroll {
+                    try await SmileID.api.doSmartSelfieEnrollment(
+                        signature: authResponse.signature,
+                        timestamp: authResponse.signature,
```
Comment (vanshg) [inline Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift:334]:
```suggestion
                        timestamp: authResponse.timestamp,
```

File: Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift [hunk 2]
```diff
@@ -314,20 +305,53 @@ public class SelfieViewModel: ObservableObject, ARKitSmileDelegate {
...
+                        request: MultiPartRequest(
+                            multiPartMedia: imageInfoArray,
+                            userId: userId,
+                            allowNewEnroll: allowNewEnroll,
+                            partnerParams: extraPartnerParams
+                        )
+                    ).async()
+                } else {
+                    try await SmileID.api.doSmartSelfieEnrollment(
```
Comment (vanshg) [inline Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift:343]: this else branch looks identical to the previous branch

File: Sources/SmileID/Classes/Networking/RestServiceClient.swift [hunk 3]
```diff
@@ -1,7 +1,8 @@
-import Foundation
 import Combine
+import Foundation
 
 protocol RestServiceClient {
     func send<T: Decodable>(request: RestRequest) -> AnyPublisher<T, Error>
+    func multipart(request: RestRequest) -> AnyPublisher<SmartSelfieResponse, Error>
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/RestServiceClient.swift:6]: Is it possible to make this generic? multipart should not be tied directly to SmartSelfieResponse

File: Sources/SmileID/Classes/Networking/Models/MultiPart.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    init?(withImage image: Data, forKey key: String, forName name: String) {
+        self.key = key
+        mimeType = "image/jpeg"
+        filename = name
+        data = image
+    }
+}
+
+public struct MultiPartRequest: Encodable {
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/MultiPart.swift:17]: Can you call this SmartSelfieRequest? Multipart is a networking concept and the request is an application level concept

**Review by vanshg — CHANGES_REQUESTED** [review]

File: Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift (hunk 2 above)
Comment (jumaallan) [reply Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift:343]: good catch, copy paste mistake

File: Sources/SmileID/Classes/Networking/RestServiceClient.swift (hunk 3 above)
Comment (jumaallan) [reply Sources/SmileID/Classes/Networking/RestServiceClient.swift:6]: I actually tried that first, and I ran into issues

I wanted to make it look like the send protocol

File: Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift (hunk 1 above)
Comment (vanshg) [reply Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift:339]: Bump

File: Example/SmileID/Util.swift [hunk 5]
```diff
@@ -11,7 +12,8 @@ func openUrl(_ urlString: String) {
 
 func jobResultMessageBuilder(
     jobName: String,
-    didSubmitJob: Bool?,
+    didSubmitJob: Bool? = nil,
+    apiResponse _: SmartSelfieResponse? = nil,
```
Comment (vanshg) [inline Example/SmileID/Util.swift:16]: We don't seem to be using this value in this function anywhere

File: Sources/SmileID/Classes/Networking/Models/SmartSelfieRequest.swift [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+struct SmartSelfieRequestImage: Encodable {
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/SmartSelfieRequest.swift:3]: This is a generic form data class and not specific to smartselfie. On Android, the Retrofit equivalent is called `MultipartBody.Part`. We should adopt a similar name. It should also be in its own file

File: Sources/SmileID/Classes/Networking/RestServiceClient.swift (hunk 3 above)
Comment (vanshg) [reply Sources/SmileID/Classes/Networking/RestServiceClient.swift:6]: What kind of issues? why can't we do something like `func multipart<T: Decodable>(request: RestRequest) -> AnyPublisher<T, Error>`

File: Sources/SmileID/Classes/Networking/Models/SmartSelfieRequest.swift [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    init?(withImage image: Data, forKey key: String, forName name: String) {
+        self.key = key
+        mimeType = "image/jpeg"
+        filename = name
+        data = image
+    }
+}
+
+public struct SmartSelfieRequest: Encodable {
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/SmartSelfieRequest.swift:17]: instead of creating a struct for this, can you encode them as top level parameters in the network interface, since they are each in independent form data part? This will also allow some parameters to be explicit where required. For example, userId is not allowed to be nil for SmartSelfieAuth, but here it is marked as optional

see here: https://github.com/smileidentity/android/blob/b7bade606af4630b6b58b8a76d34aafa69ab509b/lib/src/main/java/com/smileidentity/networking/SmileIDService.kt#L82-L90

**Review by vanshg — CHANGES_REQUESTED** [review]

File: Sources/SmileID/Classes/Networking/ServiceRunnable.swift [hunk 8]
```diff
@@ -144,4 +210,88 @@ extension ServiceRunnable {
             .setFailureType(to: Error.self)
             .eraseToAnyPublisher()
     }
+
+    func generateBoundary() -> String {
+        return UUID().uuidString
+    }
+
+    // swiftlint:disable line_length cyclomatic_complexity
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/ServiceRunnable.swift:218]: Why disable line_length?

File: Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift [hunk 9]
```diff
@@ -314,20 +305,59 @@ public class SelfieViewModel: ObservableObject, ARKitSmileDelegate {
...
+                                forName: liveness.lastPathComponent
+                            )
+                        }
+                        return nil
+                    }
+
+                    smartSelfieLivenessImages.append(contentsOf: livenessImageInfos.compactMap { $0 })
+                }
+                // swiftlint:disable line_length
+                guard let smartSelfieImage = smartSelfieImage, smartSelfieLivenessImages.count == numLivenessImages else {
```
Comment (vanshg) [inline Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift:333]: Instead of disabling this, you should be able to do something like the below (or similar):
```swift
guard let smartSelfieImage = smartSelfieImage, 
smartSelfieLivenessImages.count == numLivenessImages else {
```

**Comment by vanshg** [comment]
Please also update the CHANGELOG

File: Sources/SmileID/Classes/Networking/ServiceRunnable.swift (hunk 8 above)
Comment (jumaallan) [reply Sources/SmileID/Classes/Networking/ServiceRunnable.swift:218]: The append functions look easier to read without line breaks, but I can remove the rule and add line breaks

File: Sources/SmileID/Classes/Networking/Models/SmartSelfieRequest.swift (hunk 6 above)
Comment (jumaallan) [reply Sources/SmileID/Classes/Networking/Models/SmartSelfieRequest.swift:3]: Fixed

File: Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift [hunk 10]
```diff
@@ -314,20 +305,59 @@ public class SelfieViewModel: ObservableObject, ARKitSmileDelegate {
...
+                        return nil
+                    }
+
+                    smartSelfieLivenessImages.append(contentsOf: livenessImageInfos.compactMap { $0 })
+                }
+                guard let smartSelfieImage = smartSelfieImage,
+                smartSelfieLivenessImages.count == numLivenessImages else {
+                    throw SmileIDError.unknown("Selfie capture failed")
+                }
+                let response = if allowNewEnroll {
```
Comment (vanshg) [inline Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift:336]: Auth vs Enroll is controlled by `isEnroll`, not `allowNewEnroll`

```suggestion
                let response = if isEnroll {
```

File: Sources/SmileID/Classes/Networking/SmileIDService.swift [hunk 11]
```diff
@@ -192,6 +219,54 @@ public class SmileIDService: SmileIDServiceable, ServiceRunnable {
         post(to: "upload", with: request)
     }
 
+    public func doSmartSelfieEnrollment(
+        signature: String,
+        timestamp: String,
+        selfieImage: MultipartBody,
+        livenessImages: [MultipartBody],
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/SmileIDService.swift:226]: Can we make this `Data` and `[Data]`, and do the conversion to the MultipartBody in this method body? That way the callers don't need to work about this detail

**Review by vanshg — CHANGES_REQUESTED** [review]

File: Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift (hunk 10 above)
Comment (jumaallan) [reply Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift:336]: Good catch, fixed

File: Sources/SmileID/Classes/Networking/Models/v2/SmartSelfieResponse.swift [hunk 12]
```diff
@@ -1,14 +1,14 @@
 import Foundation
 
 public struct SmartSelfieResponse: Codable {
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/v2/SmartSelfieResponse.swift:3]: You're missing CodingKeys

```swift
public struct SmartSelfieResponse: Codable {
    public let code: String
    public let createdAt: String
    public let jobId: String
    public let jobType: JobTypeV2
    public let message: String
    public let partnerId: String
    public let partnerParams: [String: String]
    public let status: SmartSelfieStatus
    public let updatedAt: String
    public let userId: String

    enum CodingKeys: String, CodingKey {
        case code
        case createdAt = "created_at"
        case jobId = "job_id"
        case jobType = "job_type"
        case message
        case partnerId = "partner_id"
        case partnerParams = "partner_params"
        case status
        case updatedAt = "updated_at"
        case userId = "user_id"
    }
}
```

**Comment by vanshg** [comment]
We're missing the Source SDK and Version

File: Sources/SmileID/Classes/Helpers/LocalStorage.swift [hunk 13]
```diff
@@ -264,6 +264,9 @@ public class LocalStorage {
         try createDirectory(at: submittedJobDirectory)
         let unsubmittedFileDirectory = try unsubmittedJobDirectory.appendingPathComponent(jobId)
         let submittedFileDirectory = try submittedJobDirectory.appendingPathComponent(jobId)
+        if fileManager.fileExists(atPath: submittedFileDirectory.relativePath) {
+            try fileManager.removeItem(atPath: submittedFileDirectory.relativePath)
```
Comment (vanshg) [inline Sources/SmileID/Classes/Helpers/LocalStorage.swift:268]: Can you add a log here, since it's a destructive action?

Comment (jumaallan) [reply Sources/SmileID/Classes/Helpers/LocalStorage.swift:268]: the func "throws" so that would be printed where we call this `moveToSubmittedJobs` func

I don't think we need an extra log here

**Review by vanshg — APPROVED** [review]
